File: web/src/beta/fitExplanation.ts
import type { FitExplanation } from "./types";

const EVIDENCE_PREFIXES = ["Job posting:", "Confirmed information:", "Additional complete evidence"];
const UNCERTAINTY_MARKERS = [
  "unvalidated", "could not be validated", "needs review", "follow-up", "provisional",
  "unknown", "ineligible", "conflict", "not independently verified", "confirm career",
  "resolve the returned", "mandatory requirement", "not been evaluated",
  "requirement comparison was discarded", "hard constraint",
];

/** Drop internal source ids like [career_text.0] from user-visible fit copy. */
export function stripSourceCitations(text: string): string {
  return text
    .replace(/\s*\[(?:[a-z][\w]*(?:\.[a-z\d_]+)*)(?:,\s*(?:[a-z][\w]*(?:\.[a-z\d_]+)*))*\]/gi, "")
    .replace(/\s{2,}/g, " ")
    .trim();
}

function cleanLines(values: unknown, limit = 16): string[] {
  if (!Array.isArray(values)) return [];
  const items: string[] = [];
  for (const value of values.slice(0, limit)) {
    if (typeof value !== "string") continue;
    const text = stripSourceCitations(value.replace(/\s+/g, " ").trim()).slice(0, 2000);
    if (text && !items.includes(text)) items.push(text);
  }
  return items;
}

export function isFitExplanation(value: unknown): value is FitExplanation {
  if (!value || typeof value !== "object") return false;
  const item = value as FitExplanation;
  return [item.why, item.evidence, item.uncertainty].every((field) => Array.isArray(field));
}

export function fitExplanationFromRationale(rationale: string | null | undefined): FitExplanation {
  const why: string[] = [];
  const evidence: string[] = [];
  const uncertainty: string[] = [];
  for (const raw of (rationale || "").split("\n")) {
    const text = stripSourceCitations(raw.replace(/\s+/g, " ").trim());
    if (!text) continue;
    if (EVIDENCE_PREFIXES.some((prefix) => text.startsWith(prefix))) evidence.push(text);
    else if (UNCERTAINTY_MARKERS.some((marker) => text.toLowerCase().includes(marker))) uncertainty.push(text);
    else why.push(text);
  }
  return { why: cleanLines(why, 12), evidence: cleanLines(evidence, 16), uncertainty: cleanLines(uncertainty, 16) };
}

export function coerceFitExplanation(job: { rationale?: string | null; fit_explanation?: FitExplanation | null }): FitExplanation {
  const structured = job.fit_explanation;
  if (isFitExplanation(structured) && (structured.why.length || structured.evidence.length || structured.uncertainty.length)) {
    return {
      why: cleanLines(structured.why, 12),
      evidence: cleanLines(structured.evidence, 16),
      uncertainty: cleanLines(structured.uncertainty, 16),
    };
  }
  return fitExplanationFromRationale(job.rationale);
}

File: src/jobagent/mobile/fit_explanation.py
"""Additive why / evidence / uncertainty shape for ranking and assessments.

The model still selects verbatim source_facts; this module only structures the
server-built rationale. Old clients keep `rationale`. New clients may read
`fit_explanation`. Invalid stored JSON is ignored and rebuilt from rationale.
"""

from __future__ import annotations

import re
from typing import Any, Iterable, Mapping, Optional, Sequence

_MAX_ITEMS = 16
_MAX_CHARS = 2000
_SOURCE_CITATION = re.compile(
    r"\s*\[(?:[a-z][\w]*(?:\.[a-z\d_]+)*)(?:,\s*(?:[a-z][\w]*(?:\.[a-z\d_]+)*))*\]",
    re.I,
)
_EVIDENCE_PREFIXES = (
    "Job posting:",
    "Confirmed information:",
    "Additional complete evidence",
)
_UNCERTAINTY_MARKERS = (
    "unvalidated",
    "could not be validated",
    "needs review",
    "follow-up",
    "provisional",
    "unknown",
    "ineligible",
    "conflict",
    "not independently verified",
    "confirm career",
    "resolve the returned",
    "mandatory requirement",
    "not been evaluated",
    "requirement comparison was discarded",
    "hard constraint",
    "do you want to keep",
)


def _clean_line(value: Any) -> Optional[str]:
    if not isinstance(value, str):
        return None
    text = " ".join(_SOURCE_CITATION.sub("", value).split()).strip()
    if not text or len(text) > _MAX_CHARS:
        return text[:_MAX_CHARS] if text else None
    return text


def _clean_list(values: Any, *, limit: int = _MAX_ITEMS) -> list[str]:
    if not isinstance(values, list):
        return []
    items = []
    for value in values[:limit]:
        text = _clean_line(value)
        if text and text not in items:
            items.append(text)
    return items


def explanation_shape(why: Iterable[str] = (), evidence: Iterable[str] = (),
                      uncertainty: Iterable[str] = ()) -> dict:
    return {
        "why": _clean_list(list(why)),
        "evidence": _clean_list(list(evidence)),
        "uncertainty": _clean_list(list(uncertainty)),
    }


def rationale_from_entries(entries: Sequence[tuple[str, str]]) -> str:
    lines = []
    total = 0
    for _bucket, text in entries:
        cleaned = _clean_line(text)
        if not cleaned:
            continue
        if total + len(cleaned) + 1 > 8000:
            break
        lines.append(cleaned)
        total += len(cleaned) + 1
    return "\n".join(lines)


def explanation_from_entries(entries: Sequence[tuple[str, str]]) -> dict:
    buckets = {"why": [], "evidence": [], "uncertainty": []}
    for bucket, text in entries:
        if bucket not in buckets:
            bucket = "why"
        cleaned = _clean_line(text)
        if cleaned:
            buckets[bucket].append(cleaned)
    return explanation_shape(buckets["why"], buckets["evidence"], buckets["uncertainty"])


def is_evidence_line(text: str) -> bool:
    return text.startswith(_EVIDENCE_PREFIXES)


def is_uncertainty_line(text: str) -> bool:
    lowered = text.casefold()
    return any(marker in lowered for marker in _UNCERTAINTY_MARKERS)


def fit_explanation_from_rationale(rationale: Any) -> dict:
    """Backward-compatible split of a legacy concatenated rationale."""
    if not isinstance(rationale, str) or not rationale.strip():
        return explanation_shape()
    why, evidence, uncertainty = [], [], []
    for raw in rationale.splitlines():
        text = _clean_line(raw)
        if not text:
            continue
        if is_evidence_line(text):
            evidence.append(text)
        elif is_uncertainty_line(text):
            uncertainty.append(text)
        else:
            why.append(text)
    return explanation_shape(why, evidence, uncertainty)


def validated_fit_explanation(value: Any) -> Optional[dict]:
    if not isinstance(value, Mapping):
        return None
    extra = set(value) - {"why", "evidence", "uncertainty"}
    if extra:
        # Additive unknown keys are ignored, never treated as a parse failure.
        value = {key: value[key] for key in ("why", "evidence", "uncertainty") if key in value}
    shape = explanation_shape(value.get("why"), value.get("evidence"), value.get("uncertainty"))
    if not any(shape.values()):
        return None
    return shape


def coerce_fit_explanation(value: Any, *, rationale: Any = None) -> dict:
    structured = validated_fit_explanation(value)
    if structured:
        return structured
    return fit_explanation_from_rationale(rationale)
